lightningd/invoice.c

```diff
@@ -74,32 +74,44 @@ struct invoices *invoices_init(const tal_t *ctx)
 	return invs;
 }
 
+static void json_add_invoice(struct json_result *response,
+			     const struct invoice *inv)
+{
+	json_object_start(response, NULL);
+	json_add_string(response, "label", inv->label);
+	json_add_hex(response, "payment_hash", &inv->rhash, sizeof(inv->rhash));
+	if (inv->msatoshi)
+		json_add_u64(response, "msatoshi", *inv->msatoshi);
+	json_add_bool(response, "complete", inv->state == PAID);
+	if (inv->state == PAID) {
+		json_add_u64(response, "pay_index", inv->pay_index);
+		json_add_u64(response, "msatoshi_received",
+			     inv->msatoshi_received);
+	}
+	json_add_u64(response, "expiry_time", inv->expiry_time);
+	json_object_end(response);
+}
+
 static void tell_waiter(struct command *cmd, const struct invoice *paid)
 {
 	struct json_result *response = new_json_result(cmd);
 
-	json_object_start(response, NULL);
-	json_add_string(response, "label", paid->label);
-	json_add_hex(response, "rhash", &paid->rhash, sizeof(paid->rhash));
-	if (paid->msatoshi)
-		json_add_u64(response, "msatoshi", *paid->msatoshi);
-	json_add_bool(response, "complete", paid->state == PAID);
-	if (paid->state == PAID)
-		json_add_u64(response, "pay_index", paid->pay_index);
-	json_object_end(response);
+	json_add_invoice(response, paid);
 	command_success(cmd, response);
 }
 static void tell_waiter_deleted(struct command *cmd, const struct invoice *paid)
 {
 	command_fail(cmd, "invoice deleted during wait");
 }
 
-void resolve_invoice(struct lightningd *ld, struct invoice *invoice)
+void resolve_invoice(struct lightningd *ld, struct invoice *invoice,
+		     u64 msatoshi_received)
 {
 	struct invoice_waiter *w;
 	struct invoices *invs = ld->invoices;
 
 	invoice->state = PAID;
+	invoice->msatoshi_received = msatoshi_received;
```

Comment:
You load `msatoshi_received` here, but do not seem to save it in the db? Unless I mistake something.

I recommend restructuring the wallet invoice interface first precisely because of this issue. `resolve_invoice` should really access db directly since it must manipulate two columns. Non-wallet code should not be creating, initializing, and writing to invoice data structure.

Reply:
I actually prefer this way of modifying a struct and then saving it with a dumb wallet function that doesn't really know about the business logic.

Reply:
Regardless, `msatoshi_received` is not saved to the database in the code as-is.  You have `SELECT ... msatoshi_received ... FROM invoices ...` but no `UPDATE TABLE invoices SET msatoshi_received = ? ...` or `INSERT INTO TABLE invoices(... msatoshi_received ...) VALUES(...)`.

Reply:
Even the "doesn't really know the business logic" is violated, since `wallet_invoice_save` knows enough of the business logic to know that only `state` is changed, and that if `state` changes from `UNPAID` to `PAID`, it also has to generate a new `pay_index`, and (missing in this commit) also save `msatoshi_received`.  We might as well just carve the joints of the invoices system differently and expose high-level functions that directly perform database operations.  If we want to have a separate layer with a dumb database-save function, the database-save function should really save all fields indiscriminately (and as such you probably want to still have a high-level interface to wrap the database-save function).

Reply:
Oh I see, you're right, I assumed it'd be covered by the call to wallet_save_invoice on [line 118](https://github.com/rustyrussell/lightning/blob/1d0964e4ac399f909903e5550d0cb0923a3dcd0d/lightningd/invoice.c#L118). I'll amend `wallet_save_invoice` to take care of that.